File: bbtest/steps/zmq_steps.rb

step "lake recieves :data" do |data|
  ZMQHelper.send(data)
end

step "lake responds with :data" do |data|
  eventually(backoff: 0.5) {
    ok = ZMQHelper.pulled_message?(data)
    expect(ok).to be(true), "message #{data} was not found in #{ZMQHelper.mailbox()}"
  }
  ZMQHelper.ack(data)
end

step "lake performs handshake" do ||
  eventually(backoff: 0.5) {
    ZMQHelper.send("!")
    ok = ZMQHelper.pulled_message?("!")
    expect(ok).to be(true), "handshake was not relayed"
  }
  ZMQHelper.ack("!")
end

File: bbtest/steps/zmq_steps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from behave import *
from helpers.eventually import eventually
import time


@when('lake recieves "{data}"')
def lake_recieves(context, data):
  time.sleep(0.5)
  context.zmq.send(data)


@then('lake responds with "{data}"')
def lake_responds_with(context, data):
  pivot = data.encode('utf-8')
  @eventually(5)
  def impl():
    assert pivot in context.zmq.backlog, "{} not found in zmq backlog {}".format(pivot, context.zmq.backlog)
    context.zmq.ack(pivot)
  impl()
